extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -538,6 +564,161 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
     }
   }
 
+  /**
+   * Sends RTC request as specified by rtcConfig. Returns promise which
+   * resolves to the time that the RTC callout took to complete.
+   * If disableStalewhilerevalidate == true, then we will only
+   * send one RTC request per page. If it's !== true, then we always
+   * send two requests. The first will try
+   * to hit browser cache, and the second request will always bypass.
+   * This is a way for us to implement a simple stale-while-revalidate
+   * model.
+   * The targeting info from the RTC updates the targeting info on
+   * this object within mergeRtc.
+   * @return {?Promise<?Object>} An object of parameters to add to
+   *   the ad request url.
+   * @private
+   */
+  executeRtc_() {
+    if (rtcPromise) {
+      return this.mergeRtc();
+    }
+    const ampRtcPageElement = document.getElementById('amp-rtc');
+    if (!ampRtcPageElement) {
+      return Promise.resolve();
+    }
+    let endpoint;
+    rtcConfig = tryParseJson(ampRtcPageElement.textContent);
+    if (!isObject(rtcConfig) || !(endpoint = rtcConfig['endpoint']) ||
+        typeof endpoint != 'string' || !isSecureUrl(endpoint)) {
+      user().warn(TAG,
+          `invalid RTC config: ${ampRtcPageElement.textContent}`);
+      return Promise.resolve();
+    }
+    let rtcTotalTime;
+    const disableSWR = (
+        rtcConfig['disableStaleWhileRevalidate'] === true);
+    const startTime = Date.now();
+    // Because we are wrapping the RTC request in the timeout,
+    // we are guaranteeing that if the RTC is slow to return and
+    // times out for the first slot, it won't be used for any
+    // other slots either. This is in opposition to the other way
+    // we could potentially do it, in which we only still make one
+    // rtc callout, but there is a 1 second timeout from the
+    // perspective of each slot. I.e. if the top slot on the page
+    // calls out for the RTC, which returns after 5 seconds, and
+    // then a slot way down on the page asks for it later.
+    rtcPromise = Services.timerFor(window).timeoutPromise(
+        RTC_TIMEOUT,
+        Services.xhrFor(this.win).fetchJson(
+            endpoint, {credentials: 'include'}).then(res => {
+              rtcTotalTime = Date.now() - startTime;
+              if (!disableSWR) {
```

Comment:
Add a comment that SWR should be disabled if the RTC is not cachable (no cache header).  Can we read the cache header from the response?  Also something about Cache-Control not being supported in IE?